src/envelope.rs

```diff
@@ -0,0 +1,160 @@
+use crate::{Actor, AsyncHandler, Context, Handler, Message};
+use futures::channel::oneshot::{self, Receiver, Sender};
+use futures::{future, Future, FutureExt};
+use std::marker::PhantomData;
+use std::pin::Pin;
+
+type Fut<'a> = Pin<Box<dyn Future<Output = ()> + Send + 'a>>;
+
+pub(crate) trait Envelope: Send {
```

Comment:
Envelopes always return a future, does there need to be distinction between sync and async envelopes?

Reply:
Yes, it's about the `impl Envelope` thing here. If you want you can try use specialisation to make one envelope type (that can come in another PR). But I tried it and it's painful.